src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                    // is lkely going to sit around for a while in a wait loop (and the more requests
+                    // are concurrently issued to the same agent, the longer the thread will be around).
+                    const TaskCreationOptions Options = TaskCreationOptions.DenyChildAttach | TaskCreationOptions.LongRunning;
+                    Task.Factory.StartNew(s =>
+                    {
+                        var thisRef = (MultiAgent)s;
+                        try
+                        {
+                            // Do the actual processing
+                            thisRef.Process();
+                        }
+                        catch (Exception exc)
+                        {
+                            Debug.Fail("Unexpected exception from processing loop: " + exc.ToString());
+                        }
+                        finally
+                        {
+                            lock (thisRef._incomingRequests)
+                            {
+                                // Close our wakeup pipe (ignore close errors).
+                                // This is done while holding the lock to prevent
+                                // subsequent Queue calls to see an improperly configured
+                                // set of descriptors.
+                                Interop.Sys.Close(thisRef._wakeupRequestedPipeFd);
+                                Interop.Sys.Close(thisRef._requestWakeupPipeFd);
+                                thisRef._wakeupRequestedPipeFd = 0;
+                                thisRef._requestWakeupPipeFd = 0;
+
+                                // In the time between we stopped processing and now,
+                                // more requests could have been added.  If they were
+                                // kick off another processing loop.
+                                thisRef._workerRunning = false;
+                                if (thisRef._incomingRequests.Count > 0)
+                                {
+                                    thisRef.EnsureWorkerIsRunning();
+                                }
+                            }
+                        }
+                    }, this, CancellationToken.None, Options, TaskScheduler.Default);
+                }
+                else // _workerRunning == true
+                {
+                    // The worker is already running.  If there are already queued requests, we're done.
+                    // However, if there aren't any queued requests, Process could be blocked inside of
+                    // curl_multi_wait, and we want to make sure it wakes up to see that there additional
+                    // requests waiting to be handled.  So we write to the wakeup pipe.
+                    Debug.Assert(_incomingRequests.Count >= 1, "We just queued a request, so the count should be at least 1");
+                    if (_incomingRequests.Count == 1)
```

Comment:
Don't we need this for other transitions as well (eg. 1 to 2) so that curl_multi_wait updates its polling set